src/backend/cuda/memory.cpp

```diff
@@ -61,29 +57,37 @@ void printMemInfo(const char *msg, const int device) {
 
 template<typename T>
 uptr<T> memAlloc(const size_t &elements) {
-    size_t size                = elements * sizeof(T);
-    MemoryEventPair me = memoryManager().alloc(size, false);
-    cudaStream_t stream        = getActiveStream();
-    if (me.e) me.e.enqueueWait(stream);
-    return uptr<T>(static_cast<T *>(me.ptr), memFree<T>);
+    size_t size            = elements * sizeof(T);
+    af_buffer_info pair    = memoryManager().alloc(elements * sizeof(T), false);
+    BufferInfo &bufferInfo = getBufferInfo(pair);
+    detail::Event &e       = getEvent(bufferInfo.event);
```

Comment:
Restarting the converstion from [here](https://github.com/arrayfire/arrayfire/pull/2461/files/d5f3447d7e968a94f8d440fb3100c2a847cf47f0#r311183446)

> Agree that the move ctor for `af_event` can do this, but I don't see how it could work for `detail::Event`, since the `af_event` contains a pointer to heap allocated `detail::Event`, and we'd need to move that to a `detail::Event` onto the stack, which we can't do since we have no copy ctor for `detail::Event` (maybe I'm still missing something?)
> 
> > We specify that the buffer_info returned by the alloc call cannot be referenced again
> 
> Agree; if I understand correctly, you're saying we stipulate that the memory manager can't retain even referential ownership of an `af_buffer_info` after `alloc` returns. This is reasonable/seems required since we can't be sure of the author of a memory manager will free an `af_buffer_info`, so we'd be stuck in a weird spot trying to prevent a leak
> 
> Seems more obvious, but we can also specify the same thing for `unlock`: it takes an `af_event`, but it's responsible for releasing it itself.


It should work for detail::Event and by extension common::EventBase as well. The move constructor for EventBase will transfer ownership of the EventBase::EventType object that is located on the heap to the newly created EventBase on the stack. The move constructor zeros the EventBase::Event type member so the destructor doesn't call NativeEventPolicy::destroyEvent on `e_`. The heap allocated EventBase can then be deleted without affecting the underlying event and no copy is created.

This is what is going on with the Event created on the stack [here](https://github.com/arrayfire/arrayfire/pull/2461/files#diff-de07c01c5a67fbcaa47b38c3a8d3a5c6L114)

We are moving the stack allocated Event to another stack. The Event object on the current stack will still call its destructor but it will not delete the EventBase::EventType object it was holding before the move.

Reply:
Yep — you're right. I was afraid there'd be some sort of contention if we moved `e_` then freed the `af_buffer_info`, but that move ctor does the right thing as you said. Just sent that change.